src/main/scala/mesosphere/marathon/core/storage/repository/impl/CachedTemplateRepository.scala

```diff
@@ -0,0 +1,91 @@
+package mesosphere.marathon
+package core.storage.repository.impl
+
+import akka.Done
+import akka.stream.ActorMaterializer
+import com.github.blemale.scaffeine.{AsyncLoadingCache, Scaffeine}
+import mesosphere.marathon.core.storage.repository.TemplateRepositoryLike.{Spec, Template}
+import mesosphere.marathon.core.storage.zookeeper.ZooKeeperPersistenceStore
+import mesosphere.marathon.state.PathId
+
+import scala.async.Async._
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+  * This is a cached version of the [[TemplateRepository]]. It uses [[https://github.com/blemale/scaffeine Scaffeine library]]
+  * which itself is a Scala wrapper over the [[https://github.com/ben-manes/caffeine Caffeine cache]]. It supports asynchronous
+  * loading of cache elements, size-based eviction, time-based expiration of elements etc.
+  *
+  * All template versions are saved in the cache using [[CachedTemplateRepository#toCacheKey()]]
+  * method e.g. a template with pathId=/foo/bar and version=1 will be saved using the key `foo_bar:1`.
+  *
+  * Basically this class wraps [[TemplateRepository]] methods with the calls to cache. [[read()]] method goes directly to cache
+  * and the element will be loaded if not already exists.
+  *
+  * Note that this implementation is tread-safe and provides serializability guarantees e.g. two independent [[create()]]
+  * calls are not guaranteed to be executed in any particular order the same way one should not expect two independent futures
+  * to have an order.
+  *
+  * Known pitfalls: an operation e.g. [[delete()]] may fail leaving the cache and the underlying store in an inconsistent
```

Comment:
Is it possible to workaround this inside the implementation so that it is hidden from a user?

Reply:
I was referring to a specific use case where an error occurred during store's `delete` operation, meaning that _most probably_ the ZK connection is gone and this situation will be handled on the "higher application level". Otherwise, I don't see how to meaningfully handle this kind of errors on this level.